src/repr/src/adt/range.rs

```diff
@@ -0,0 +1,329 @@
+    pub inner: Option<RangeInner<'a>>,
+}
+
+impl<'a> std::fmt::Display for Range<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self.inner {
+            None => f.write_str("empty"),
+            Some(i) => write!(f, "{}", i),
+        }
+    }
+}
+
+/// Holds the upper and lower `DatumRangeBound`s.
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash)]
+pub struct RangeInner<'a> {
+    pub lower: RangeBound<'a>,
+    pub upper: RangeBound<'a>,
+}
+
+impl<'a> std::fmt::Display for RangeInner<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(if self.lower.inclusive { "[" } else { "(" })?;
+        write!(f, "{}", self.lower)?;
+        f.write_str(",")?;
+        write!(f, "{}", self.upper)?;
+        f.write_str(if self.upper.inclusive { "]" } else { ")" })
+    }
+}
+
+impl<'a> Ord for RangeInner<'a> {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        match self.lower.cmp(&other.lower) {
+            std::cmp::Ordering::Equal => self.upper.cmp(&other.upper).reverse(),
+            o => o,
+        }
+    }
+}
+
+impl<'a> PartialOrd for RangeInner<'a> {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+/// Represents a terminal point of a range.
+///
+/// Note that this struct's implementation of `Ord` is complex, and should
+/// probably not be used directly; instead, prefer comparing `RangeInner`.
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash)]
+pub struct RangeBound<'a> {
+    pub inclusive: bool,
+    pub bound: Option<DatumNested<'a>>,
+}
+
+impl<'a> std::fmt::Display for RangeBound<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self.bound {
+            None => Ok(()),
+            Some(bound) => write!(f, "{}", bound.datum()),
+        }
+    }
+}
+
+impl<'a> Ord for RangeBound<'a> {
+    /// `cmp` is only correct comparing lower bounds.
+    ///
+    /// - The output is correct for upper bounds if `Ordering::reverse`'d.
+    /// - Does not support comparing an upper and a lower bound.
+    fn cmp(&self, other: &Self) -> Ordering {
+        match self.bound.cmp(&other.bound) {
+            Ordering::Equal => {
+                if self.inclusive == other.inclusive {
+                    Ordering::Equal
+                } else if self.inclusive {
+                    Ordering::Less
+                } else {
+                    Ordering::Greater
+                }
+            }
+            o => o,
+        }
+    }
+}
+
+impl<'a> PartialOrd for RangeBound<'a> {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+/// A contrivance to make arguments to functions that construct ranges more
+/// well structured.
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash, PartialOrd)]
+pub struct RangeBoundDesc {
+    // Is this bound infinite/unbounded?
+    pub inf: bool,
```

Comment:
(See comment elsewhere about it seeming nice to just always spell out these terms.)

```suggestion
    pub infinite: bool,
```